compiler/cpp/src/thrift/generate/t_go_generator.cc

```diff
@@ -2749,150 +2754,184 @@ void t_go_generator::generate_service_server(t_service* tservice) {
+    f_types_ << indent() << "}" << endl << endl;
+    f_types_ << indent() << "iprot.ReadMessageEnd()" << endl;
 
-  if (!tfunction->is_oneway()) {
-    f_types_ << indent() << "result := " << resultname << "{}" << endl;
-  }
-  bool need_reference = type_need_reference(tfunction->get_returntype());
-  if (!tfunction->is_oneway() && !tfunction->get_returntype()->is_void()) {
-    f_types_ << "var retval " << type_to_go_type(tfunction->get_returntype()) << endl;
-  }
+    if (!tfunction->is_oneway()) {
+        f_types_ << indent() << "result := " << resultname << "{}" << endl;
+    }
+    bool need_reference = type_need_reference(tfunction->get_returntype());
+    if (!tfunction->is_oneway() && !tfunction->get_returntype()->is_void()) {
+        f_types_ << indent() << "var retval " << type_to_go_type(tfunction->get_returntype()) << endl;
+    }
 
-  f_types_ << indent() << "var err2 error" << endl;
-  f_types_ << indent() << "if ";
+    f_types_ << indent() << "var err2 error" << endl;
+    f_types_ << indent();
+    if (!tfunction->is_oneway()) {
+        if (!tfunction->get_returntype()->is_void()) {
+            f_types_ << "res, ";
+        } else {
+            f_types_ << "_, ";
+        }
+    } else {
+        f_types_ << "_, ";
+    }
 
-  if (!tfunction->is_oneway()) {
-    if (!tfunction->get_returntype()->is_void()) {
-      f_types_ << "retval, ";
+    // Generate the function call
+    t_struct* arg_struct = tfunction->get_arglist();
+    const std::vector<t_field*>& fields = arg_struct->get_members();
+    vector<t_field*>::const_iterator f_iter;
+
+    f_types_ << "err2 ";
+
+    if (!tfunction->is_oneway()) {
+        if (!tfunction->get_returntype()->is_void()) {
+            f_types_ << ":";
+        }
     }
-  }
 
-  // Generate the function call
-  t_struct* arg_struct = tfunction->get_arglist();
-  const std::vector<t_field*>& fields = arg_struct->get_members();
-  vector<t_field*>::const_iterator f_iter;
-  f_types_ << "err2 = p.handler." << publicize(tfunction->get_name()) << "(";
-  bool first = true;
+    f_types_ << "= p.opts.Interceptor(ctx, \"" << escape_string(tfunction->get_name()) << "\", args, func(ctx context.Context, arg interface{}) (result interface{}, err error) {" << endl;
+    f_types_ << indent() << indent() << "return ";
 
-  f_types_ << "ctx";
-  for (f_iter = fields.begin(); f_iter != fields.end(); ++f_iter) {
-    if (first) {
-      first = false;
-      f_types_ << ", ";
-    } else {
-      f_types_ << ", ";
+    if (tfunction->get_returntype()->is_void()) {
+        f_types_ << "nil, ";
     }
 
-    f_types_ << "args." << publicize((*f_iter)->get_name());
-  }
+    f_types_ << "p.handler." << publicize(tfunction->get_name()) << "(";
+    bool first = true;
+
+    f_types_ << "ctx";
+    for (f_iter = fields.begin(); f_iter != fields.end(); ++f_iter) {
+        if (first) {
+            first = false;
+            f_types_ << ", ";
+        } else {
+            f_types_ << ", ";
+        }
```

Comment:
can be optimized:

        if (first) {
            first = false;
            f_types_ << ", ";
        } else {
            f_types_ << ", ";
        } 

simply

            f_types_ << ", ";

and the first seems not needed anymore at all.